### ISSUE TITLE
React v0.14.5: setState doesn't work for Shallow Render

### ISSUE THREAD

**Issue by alexeyraspopov** [facebook/react#5779]
You can use [this gist](https://gist.github.com/alexeyraspopov/3c021a41180c176130fb) for reproducing the issue.

For the 1st test:

I want to make an async request when component is going to be mounted and use the result of a request for updating state. As you will see in test report, it doesn't happen.

For the 2nd test:

I believe this issue was fixed recently, in 0.14.3. Shallow Render can't recognize sync state updates.

Please tell me if you see any mistakes in my code, because I was trying to fix these problems in many ways. I hope example code is simple enough.

**Comment by jimfb** [comment]
These sound like usage questions, rather than a bug in the React core. Usage questions are better answered on sites like StackOverflow, as we try to use github issues for tracking bugs in the React core. For this reason, I'm going to close out this issue, but feel free to continue the conversation here or move it to StackOverflow.

**Comment by alexeyraspopov** [comment]
Okay, so here is logs.

## React v0.14.3

```
~/oss/3c021a41180c176130fb master*
λ npm install react@0.14.3
npm WARN package.json issue@ No repository field.
npm WARN package.json issue@ No license field.
react@0.14.3 node_modules/react
├── envify@3.4.0 (through@2.3.8, jstransform@10.1.0)
└── fbjs@0.3.2 (whatwg-fetch@0.9.0, ua-parser-js@0.7.10, loose-envify@1.1.0, promise@7.1.1, core-js@1.2.6)
~/oss/3c021a41180c176130fb master*
λ npm test

> issue@ test /Users/alexey/oss/3c021a41180c176130fb
> mocha --compilers js:babel/register index.js



  React Shallow Render Issues
    ✓ should update state when promise is resolved
    1) should update state immediately


  1 passing (77ms)
  1 failing

  1) React Shallow Render Issues should update state immediately:
     TypeError: Cannot read property '_currentElement' of null
      at [object Object].ReactCompositeComponentMixin._updateRenderedComponent (node_modules/react/lib/ReactCompositeComponent.js:559:52)
      at [object Object].ReactCompositeComponentMixin._performComponentUpdate (node_modules/react/lib/ReactCompositeComponent.js:544:10)
      at [object Object].ReactCompositeComponentMixin.updateComponent (node_modules/react/lib/ReactCompositeComponent.js:473:12)
      at [object Object].wrapper [as updateComponent] (node_modules/react/lib/ReactPerf.js:66:21)
      at [object Object].ReactCompositeComponentMixin.performUpdateIfNecessary (node_modules/react/lib/ReactCompositeComponent.js:421:12)
      at Object.ReactReconciler.performUpdateIfNecessary (node_modules/react/lib/ReactReconciler.js:102:22)
      at runBatchedUpdates (node_modules/react/lib/ReactUpdates.js:129:21)
      at ReactReconcileTransaction.Mixin.perform (node_modules/react/lib/Transaction.js:136:20)
      at ReactUpdatesFlushTransaction.Mixin.perform (node_modules/react/lib/Transaction.js:136:20)
      at ReactUpdatesFlushTransaction.assign.perform (node_modules/react/lib/ReactUpdates.js:86:38)
      at Object.flushBatchedUpdates (node_modules/react/lib/ReactUpdates.js:147:19)
      at Object.wrapper [as flushBatchedUpdates] (node_modules/react/lib/ReactPerf.js:66:21)
      at ReactDefaultBatchingStrategyTransaction.Mixin.closeAll (node_modules/react/lib/Transaction.js:202:25)
      at ReactDefaultBatchingStrategyTransaction.Mixin.perform (node_modules/react/lib/Transaction.js:149:16)
      at Object.ReactDefaultBatchingStrategy.batchedUpdates (node_modules/react/lib/ReactDefaultBatchingStrategy.js:62:19)
      at Object.enqueueUpdate (node_modules/react/lib/ReactUpdates.js:176:22)
      at enqueueUpdate (node_modules/react/lib/ReactUpdateQueue.js:24:16)
      at Object.ReactUpdateQueue.enqueueSetState (node_modules/react/lib/ReactUpdateQueue.js:190:5)
      at TestComponent.ReactComponent.setState (node_modules/react/lib/ReactComponent.js:65:16)
      at TestComponent.update (index.js:44:14)
      at TestComponent.componentWillMount (index.js:48:14)
      at [object Object].ReactCompositeComponentMixin.mountComponent (node_modules/react/lib/ReactCompositeComponent.js:210:12)
      at [object Object].wrapper [as mountComponent] (node_modules/react/lib/ReactPerf.js:66:21)
      at [object Object].ReactShallowRenderer._render (node_modules/react/lib/ReactTestUtils.js:362:14)
      at [object Object].ReactShallowRenderer.render (node_modules/react/lib/ReactTestUtils.js:344:8)
      at Context.<anonymous> (index.js:57:18)



npm ERR! Test failed.  See above for more details.
```

## React v0.14.4

```
~/oss/3c021a41180c176130fb master*
λ npm install react@0.14.4
npm WARN package.json issue@ No repository field.
npm WARN package.json issue@ No license field.
react@0.14.4 node_modules/react
├── envify@3.4.0 (through@2.3.8, jstransform@10.1.0)
└── fbjs@0.3.2 (whatwg-fetch@0.9.0, ua-parser-js@0.7.10, loose-envify@1.1.0, promise@7.1.1, core-js@1.2.6)
~/oss/3c021a41180c176130fb master*
λ npm test

> issue@ test /Users/alexey/oss/3c021a41180c176130fb
> mocha --compilers js:babel/register index.js



  React Shallow Render Issues
    ✓ should update state when promise is resolved
    1) should update state immediately


  1 passing (82ms)
  1 failing

  1) React Shallow Render Issues should update state immediately:
     TypeError: Cannot read property '_currentElement' of null
      at [object Object].ReactCompositeComponentMixin._updateRenderedComponent (node_modules/react/lib/ReactCompositeComponent.js:559:52)
      at [object Object].ReactCompositeComponentMixin._performComponentUpdate (node_modules/react/lib/ReactCompositeComponent.js:544:10)
      at [object Object].ReactCompositeComponentMixin.updateComponent (node_modules/react/lib/ReactCompositeComponent.js:473:12)
      at [object Object].wrapper [as updateComponent] (node_modules/react/lib/ReactPerf.js:66:21)
      at [object Object].ReactCompositeComponentMixin.performUpdateIfNecessary (node_modules/react/lib/ReactCompositeComponent.js:421:12)
      at Object.ReactReconciler.performUpdateIfNecessary (node_modules/react/lib/ReactReconciler.js:102:22)
      at runBatchedUpdates (node_modules/react/lib/ReactUpdates.js:129:21)
      at ReactReconcileTransaction.Mixin.perform (node_modules/react/lib/Transaction.js:136:20)
      at ReactUpdatesFlushTransaction.Mixin.perform (node_modules/react/lib/Transaction.js:136:20)
      at ReactUpdatesFlushTransaction.assign.perform (node_modules/react/lib/ReactUpdates.js:86:38)
      at Object.flushBatchedUpdates (node_modules/react/lib/ReactUpdates.js:147:19)
      at Object.wrapper [as flushBatchedUpdates] (node_modules/react/lib/ReactPerf.js:66:21)
      at ReactDefaultBatchingStrategyTransaction.Mixin.closeAll (node_modules/react/lib/Transaction.js:202:25)
      at ReactDefaultBatchingStrategyTransaction.Mixin.perform (node_modules/react/lib/Transaction.js:149:16)
      at Object.ReactDefaultBatchingStrategy.batchedUpdates (node_modules/react/lib/ReactDefaultBatchingStrategy.js:62:19)
      at Object.enqueueUpdate (node_modules/react/lib/ReactUpdates.js:176:22)
      at enqueueUpdate (node_modules/react/lib/ReactUpdateQueue.js:24:16)
      at Object.ReactUpdateQueue.enqueueSetState (node_modules/react/lib/ReactUpdateQueue.js:190:5)
      at TestComponent.ReactComponent.setState (node_modules/react/lib/ReactComponent.js:65:16)
      at TestComponent.update (index.js:44:14)
      at TestComponent.componentWillMount (index.js:48:14)
      at [object Object].ReactCompositeComponentMixin.mountComponent (node_modules/react/lib/ReactCompositeComponent.js:210:12)
      at [object Object].wrapper [as mountComponent] (node_modules/react/lib/ReactPerf.js:66:21)
      at [object Object].ReactShallowRenderer._render (node_modules/react/lib/ReactTestUtils.js:362:14)
      at [object Object].ReactShallowRenderer.render (node_modules/react/lib/ReactTestUtils.js:344:8)
      at Context.<anonymous> (index.js:57:18)



npm ERR! Test failed.  See above for more details.
```

## React v0.14.5

```
~/oss/3c021a41180c176130fb master*
λ npm install react@0.14.5
npm WARN package.json issue@ No repository field.
npm WARN package.json issue@ No license field.
react@0.14.5 node_modules/react
├── envify@3.4.0 (through@2.3.8, jstransform@10.1.0)
└── fbjs@0.6.0 (whatwg-fetch@0.9.0, ua-parser-js@0.7.10, loose-envify@1.1.0, promise@7.1.1, core-js@1.2.6)
~/oss/3c021a41180c176130fb master*
λ npm test

> issue@ test /Users/alexey/oss/3c021a41180c176130fb
> mocha --compilers js:babel/register index.js



  React Shallow Render Issues
    1) should update state when promise is resolved
    2) should update state immediately


  0 passing (73ms)
  2 failing

  1) React Shallow Render Issues should update state when promise is resolved:

      AssertionError: The state should be updated
      + expected - actual

      -pending
      +success

      at index.js:32:14
      at run (node_modules/babel/node_modules/babel-core/node_modules/core-js/modules/es6.promise.js:104:47)
      at node_modules/babel/node_modules/babel-core/node_modules/core-js/modules/es6.promise.js:115:28
      at flush (node_modules/babel/node_modules/babel-core/node_modules/core-js/modules/$.microtask.js:19:5)

  2) React Shallow Render Issues should update state immediately:
     ReferenceError: document is not defined
      at getActiveElement (node_modules/react/node_modules/fbjs/lib/getActiveElement.js:25:12)
      at ReactReconcileTransaction.ReactInputSelection.getSelectionInformation (node_modules/react/lib/ReactInputSelection.js:38:23)
      at ReactReconcileTransaction.Mixin.initializeAll (node_modules/react/lib/Transaction.js:168:75)
      at ReactReconcileTransaction.Mixin.perform (node_modules/react/lib/Transaction.js:135:12)
      at ReactUpdatesFlushTransaction.Mixin.perform (node_modules/react/lib/Transaction.js:136:20)
      at ReactUpdatesFlushTransaction.assign.perform (node_modules/react/lib/ReactUpdates.js:86:38)
      at Object.flushBatchedUpdates (node_modules/react/lib/ReactUpdates.js:147:19)
      at Object.wrapper [as flushBatchedUpdates] (node_modules/react/lib/ReactPerf.js:66:21)
      at ReactDefaultBatchingStrategyTransaction.Mixin.closeAll (node_modules/react/lib/Transaction.js:202:25)
      at ReactDefaultBatchingStrategyTransaction.Mixin.perform (node_modules/react/lib/Transaction.js:149:16)
      at Object.ReactDefaultBatchingStrategy.batchedUpdates (node_modules/react/lib/ReactDefaultBatchingStrategy.js:62:19)
      at Object.enqueueUpdate (node_modules/react/lib/ReactUpdates.js:176:22)
      at enqueueUpdate (node_modules/react/lib/ReactUpdateQueue.js:24:16)
      at Object.ReactUpdateQueue.enqueueSetState (node_modules/react/lib/ReactUpdateQueue.js:190:5)
      at TestComponent.ReactComponent.setState (node_modules/react/lib/ReactComponent.js:65:16)
      at TestComponent.update (index.js:44:14)
      at TestComponent.componentWillMount (index.js:48:14)
      at [object Object].ReactCompositeComponentMixin.mountComponent (node_modules/react/lib/ReactCompositeComponent.js:210:12)
      at [object Object].wrapper [as mountComponent] (node_modules/react/lib/ReactPerf.js:66:21)
      at [object Object].ReactShallowRenderer._render (node_modules/react/lib/ReactTestUtils.js:362:14)
      at [object Object].ReactShallowRenderer.render (node_modules/react/lib/ReactTestUtils.js:344:8)
      at Context.<anonymous> (index.js:57:18)



npm ERR! Test failed.  See above for more details.
```

As you can see in `v0.14.3` and `v0.14.4` first test (with promise) works. The second test has known issue that was fixed by #5561. And after that PR has been merged (I guess it happens before `v0.14.5`) all tests become broken.

**Comment by alexeyraspopov** [comment]
As you also can see, the error message for 2nd test with `v0.14.5` is different.

**Comment by alexeyraspopov** [comment]
I might not care about 2nd because it's sync update. But 1st test is well know practice and I bet someone else will get broken tests after upgrade to `v0.14.5`.

**Comment by alexeyraspopov** [comment]
In my case, I'm using `~` for version range and I've got red build as soon as newest version has been released.

**Comment by alexeyraspopov** [comment]
Looking at changes between all these version I think it can be a problem because of `fbjs@0.6.0` that is used in `v0.14.5`.

**Comment by zpao** [comment]
Not a usage question and yea, this is getting fixed in fbjs. We'll have to publish something for React…

**Comment by MikeMcElroy** [comment]
Is this potentially related to #5561?  Can confirm that change fixed the `Cannot read property '_currentElement' of null` error for me.  It landed in master over a month ago, no idea why it hasn't peppered out in 0.14.4 or 0.14.5.

**Comment by MikeMcElroy** [comment]
Just saw that @alexeyraspopov commented about 5561.  I also just confirmed that that code is not present in [v0.14.5](https://github.com/facebook/react/blob/v0.14.5/src/test/ReactTestUtils.js#L426-L428) or [v0.14.4](https://github.com/facebook/react/blob/v0.14.4/src/test/ReactTestUtils.js#L426-L428), but is present in [master](https://github.com/facebook/react/blob/master/src/test/ReactTestUtils.js#L430-L439)

**Comment by zpao** [comment]
@MikeMcElroy no, this isn't related. #5561 has not be merged to the 0.14 release. The issue here is an accidental change in a dependency that got picked up (https://github.com/facebook/fbjs/pull/96 for the fix, just deciding on the right way to version to fix this).

**Comment by zpao** [comment]
Shipped 0.14.6 so this should get us back to the behavior we had before 0.14.4/5.

**Comment by alexeyraspopov** [comment]
```
~/oss/3c021a41180c176130fb master*
λ npm install --save react@0.14.6
npm WARN package.json issue@ No repository field.
npm WARN package.json issue@ No license field.
react@0.14.6 node_modules/react
├── envify@3.4.0 (through@2.3.8, jstransform@10.1.0)
└── fbjs@0.6.1 (whatwg-fetch@0.9.0, loose-envify@1.1.0, ua-parser-js@0.7.10, promise@7.1.1, core-js@1.2.6)
~/oss/3c021a41180c176130fb master*
λ npm test

> issue@ test /Users/alexey/oss/3c021a41180c176130fb
> mocha --compilers js:babel/register index.js



  React Shallow Render Issues
    ✓ should update state when promise is resolved
    1) should update state immediately


  1 passing (68ms)
  1 failing

  1) React Shallow Render Issues should update state immediately:
     TypeError: Cannot read property '_currentElement' of null
      at [object Object].ReactCompositeComponentMixin._updateRenderedComponent (node_modules/react/lib/ReactCompositeComponent.js:559:52)
      at [object Object].ReactCompositeComponentMixin._performComponentUpdate (node_modules/react/lib/ReactCompositeComponent.js:544:10)
      at [object Object].ReactCompositeComponentMixin.updateComponent (node_modules/react/lib/ReactCompositeComponent.js:473:12)
      at [object Object].wrapper [as updateComponent] (node_modules/react/lib/ReactPerf.js:66:21)
      at [object Object].ReactCompositeComponentMixin.performUpdateIfNecessary (node_modules/react/lib/ReactCompositeComponent.js:421:12)
      at Object.ReactReconciler.performUpdateIfNecessary (node_modules/react/lib/ReactReconciler.js:102:22)
      at runBatchedUpdates (node_modules/react/lib/ReactUpdates.js:129:21)
      at ReactReconcileTransaction.Mixin.perform (node_modules/react/lib/Transaction.js:136:20)
      at ReactUpdatesFlushTransaction.Mixin.perform (node_modules/react/lib/Transaction.js:136:20)
      at ReactUpdatesFlushTransaction.assign.perform (node_modules/react/lib/ReactUpdates.js:86:38)
      at Object.flushBatchedUpdates (node_modules/react/lib/ReactUpdates.js:147:19)
      at Object.wrapper [as flushBatchedUpdates] (node_modules/react/lib/ReactPerf.js:66:21)
      at ReactDefaultBatchingStrategyTransaction.Mixin.closeAll (node_modules/react/lib/Transaction.js:202:25)
      at ReactDefaultBatchingStrategyTransaction.Mixin.perform (node_modules/react/lib/Transaction.js:149:16)
      at Object.ReactDefaultBatchingStrategy.batchedUpdates (node_modules/react/lib/ReactDefaultBatchingStrategy.js:62:19)
      at Object.enqueueUpdate (node_modules/react/lib/ReactUpdates.js:176:22)
      at enqueueUpdate (node_modules/react/lib/ReactUpdateQueue.js:24:16)
      at Object.ReactUpdateQueue.enqueueSetState (node_modules/react/lib/ReactUpdateQueue.js:190:5)
      at TestComponent.ReactComponent.setState (node_modules/react/lib/ReactComponent.js:65:16)
      at TestComponent.update (index.js:44:14)
      at TestComponent.componentWillMount (index.js:48:14)
      at [object Object].ReactCompositeComponentMixin.mountComponent (node_modules/react/lib/ReactCompositeComponent.js:210:12)
      at [object Object].wrapper [as mountComponent] (node_modules/react/lib/ReactPerf.js:66:21)
      at [object Object].ReactShallowRenderer._render (node_modules/react/lib/ReactTestUtils.js:362:14)
      at [object Object].ReactShallowRenderer.render (node_modules/react/lib/ReactTestUtils.js:344:8)
      at Context.<anonymous> (index.js:57:18)



npm ERR! Test failed.  See above for more details.
```

Sounds right, but what about this particular bug?

**Comment by MikeMcElroy** [comment]
@alexeyraspopov that's the kind of error I get when I don't apply #5561 to my own version of React.  Could you try applying it to see if it works?

**Comment by alexeyraspopov** [comment]
@MikeMcElroy and as I understood #5561 wasn't included in `v0.14.6`. As I see in tests for this PR, it really fixes the problem. The fix just needs to be included in the next release.

**Comment by alexeyraspopov** [comment]
I'm sure it will be fixed in `v0.15` since the team will create `v0.15-stable` branch based on `master`.

**Comment by MikeMcElroy** [comment]
Is there an ETA on 0.15, or a particular reason it can't be moved into 0.14?

**Comment by zpao** [comment]
0.15 will ship when it's ready. Probably not this month. #5561 was not shipped in 0.14.x because nobody thought about it nor asked for it to be.

This bug was really about a broken 0.14.5. #5561 solves the rest of your issue. If you have concerns about #5561, please discuss there so the conversation can be correlated.
